packages/tlschema/src/createTldrawEditorSchema.ts

```diff
@@ -0,0 +1,148 @@
+import { storeMigrations } from './schema'
+import { arrowShapeTypeMigrations, arrowShapeTypeValidator } from './shapes/TLArrowShape'
+import { bookmarkShapeTypeMigrations, bookmarkShapeTypeValidator } from './shapes/TLBookmarkShape'
+import { drawShapeTypeMigrations, drawShapeTypeValidator } from './shapes/TLDrawShape'
+import { embedShapeTypeMigrations, embedShapeTypeValidator } from './shapes/TLEmbedShape'
+import { frameShapeTypeMigrations, frameShapeTypeValidator } from './shapes/TLFrameShape'
+import { geoShapeTypeMigrations, geoShapeTypeValidator } from './shapes/TLGeoShape'
+import { groupShapeTypeMigrations, groupShapeTypeValidator } from './shapes/TLGroupShape'
+import { imageShapeTypeMigrations, imageShapeTypeValidator } from './shapes/TLImageShape'
+import { lineShapeTypeMigrations, lineShapeTypeValidator } from './shapes/TLLineShape'
+import { noteShapeTypeMigrations, noteShapeTypeValidator } from './shapes/TLNoteShape'
+import { textShapeTypeMigrations, textShapeTypeValidator } from './shapes/TLTextShape'
+import { videoShapeTypeMigrations, videoShapeTypeValidator } from './shapes/TLVideoShape'
+
+/** @public */
+export type SchemaShapeInfo = {
+	migrations?: Migrations
+	validator?: { validate: (record: any) => any }
+}
+
+const defaultShapes: Record<string, SchemaShapeInfo> = {
+	arrow: {
+		migrations: arrowShapeTypeMigrations,
+		validator: arrowShapeTypeValidator,
+	},
+	bookmark: {
+		migrations: bookmarkShapeTypeMigrations,
+		validator: bookmarkShapeTypeValidator,
+	},
+	draw: {
+		migrations: drawShapeTypeMigrations,
+		validator: drawShapeTypeValidator,
+	},
+	embed: {
+		migrations: embedShapeTypeMigrations,
+		validator: embedShapeTypeValidator,
+	},
+	frame: {
+		migrations: frameShapeTypeMigrations,
+		validator: frameShapeTypeValidator,
+	},
+	geo: {
+		migrations: geoShapeTypeMigrations,
+		validator: geoShapeTypeValidator,
+	},
+	group: {
+		migrations: groupShapeTypeMigrations,
+		validator: groupShapeTypeValidator,
+	},
+	image: {
+		migrations: imageShapeTypeMigrations,
+		validator: imageShapeTypeValidator,
+	},
+	line: {
+		migrations: lineShapeTypeMigrations,
+		validator: lineShapeTypeValidator,
+	},
+	note: {
+		migrations: noteShapeTypeMigrations,
+		validator: noteShapeTypeValidator,
+	},
+	text: {
+		migrations: textShapeTypeMigrations,
+		validator: textShapeTypeValidator,
+	},
+	video: {
+		migrations: videoShapeTypeMigrations,
+		validator: videoShapeTypeValidator,
+	},
+}
+
+/**
+ * Create a TLSchema with custom shapes. Custom shapes cannot override default shapes.
+ *
+ * @param opts - Options
+ *
+ * @public */
+export function createTldrawEditorSchema(
+	opts = {} as {
+		customShapes: Record<string, SchemaShapeInfo>
+	}
+) {
+	const { customShapes } = opts
+	for (const key in customShapes) {
+		if (key in defaultShapes) {
+			throw Error(`Can't override default shape ${key}!`)
+		}
+	}
+
+	const allShapeEntries = Object.entries({ ...customShapes, ...defaultShapes })
```

Comment:
Yeah, the idea here would be to tighten things down before we open them up again. 

At the moment we don't really have a way of accounting for having our shapes overwritten. It sounded like something that could cause problems later on, or which might require special cases in how shapes are migrated / validated. 

I've updated it so that we distinguish between "core" shapes that are baked into the App's logic (e.g. groups, text that gets created on double click, or images / videos when pasting in those formats). Over time we can move these core shapes out as our references to them get reduced.